### PR TITLE
New LibraryBuilder class

### PR DESCRIPTION
**Pull request by josephmancuso**
Combines https://github.com/scriptotek/php-google-books/pull/3 and https://github.com/scriptotek/php-google-books/pull/2

### CODE REVIEW

**Comment by danmichaelo** [comment]
There is something that troubles me about the syntax here. First, I feel like the order is a bit arbitrary. I think I would easily write `$books->volumes->limit(20)->chunk(3)->search($_GET['search']);` on a bad day, but this would fail. Second thing is that it allows for nonsensical combinations like `$books->limit(20)->volumes->chunk(3)->byIsn('...');`. Thinking of an alternative approach, what if we let the `search()` method return an object of a new class (a little bit like [QueryBuilder](https://laravel.com/docs/5.5/queries)) that holds the `limit()` (or `take()`) and `chunk()` methods? This could also have a `skip()` method that sets the `startIndex` API parameter, which is useful if you want to provide multi page navigation. Do you get where I'm going?

**Comment by josephmancuso** [comment]
very true. So bascially an intermediary class where we can load it up with these option methods. I wasn't sure if you wanted to keep the generators intact which is why i kept trying to return generators. I knew the order was really weird but I kept trying to to order things differently but the seperation of GoogleBooks and Volumes is what drove me to seperate the classes and the generators are strange in nature as no other code is executed into a generator. 

for example:

```php
public function gen()
{
    $this->prop = 1;
    yield 5;
}
```

never exeutes setting the property when called. which is why i seperated the search method into 2 seperate methods.

So are you saying keep the generator for getting the API but handle it differently via a new class?

**Comment by josephmancuso** [comment]
We can make a class of something like `LibraryBuilder` or `SearchBuilder` and load it up with eloquent like methods like `find($id)`, `search()`, `filter()`, `all()`, `chunk()`. I'd like to make the API pretty intelligent and then build out the docs

**Comment by danmichaelo** [comment]
> very true. So bascially an intermediary class where we can load it up with these option methods.

Yup!

> I wasn't sure if you wanted to keep the generators intact which is why i kept trying to return generators. 

It's not really important. My only requirement is a way to easily iterate over all the results, but this can be achieved in different ways. Either making the class itself iterable, or perhaps cleaner, add a `iter()` method.

> I knew the order was really weird but I kept trying to to order things differently but the seperation of GoogleBooks and Volumes is what drove me to seperate the classes and the generators are strange in nature as no other code is executed into a generator.
>
> for example:
>
> public function gen()
> {
>    $this->prop = 1;
>    yield 5;
> }
> never exeutes setting the property when called. which is why i seperated the search method into 2 seperate methods.
> 
> So are you saying keep the generator for getting the API but handle it differently via a new class?

Not totally sure what you mean in the last question, but I'm open to dropping generators here and there. Most important is that the resulting interface is easy to work with. And I'm ok with some breaking changes, we're at version 0.2.3 after all :)

File: src/LibraryBuilder.php [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+     */
+    public function __construct(GoogleBooks $GoogleBooks)
+    {
+        $this->client = $GoogleBooks;
+    }
+
+    /**
+     * Limits the results. May be an integer between 1 and 40
+     *
+     * @param [int] $limit
```
Comment (josephmancuso) [inline src/LibraryBuilder.php]: fix doctypes

File: src/LibraryBuilder.php [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+     * Fetches results from the Google API
+     *
+     * @param [string] $id
+     * @return array
+     */
+    private function fetch($id = false)
+    {
+        if ($this->limit <= 40) {
+            if ($id) {
+                return $this->client->getItem("volumes/$id", $this->getParams());
```
Comment (josephmancuso) [inline src/LibraryBuilder.php]: remove additional parameter in getItem()

File: src/LibraryBuilder.php [hunk 3]
```diff
@@ -78,7 +78,7 @@ private function fetchAll()
     /**
      * Finds and returns one result. Should accept a book ID (found on API responses)
      *
-     * @param [string] $id
+     * @param string $id
      * @return Scriptotek\GoogleBooks\Volume;
```
Comment (josephmancuso) [inline src/LibraryBuilder.php]: remove semi-colon at end of docblock

File: src/LibraryBuilder.php [hunk 4]
```diff
@@ -114,7 +114,7 @@ public function all()
     /**
      * Sets the chunk property. Used as an alias for array chunking. Primarily for views
      *
-     * @param [int] $chunk
+     * @param int $chunk
      * @return self
      */
     public function chunk($chunk)
```
Comment (josephmancuso) [inline src/LibraryBuilder.php]: remove $chunk = false from property

**Comment by josephmancuso** [comment]
- So one of us will have to rewrite the tests for this as some are breaking changes since I introduced a `LibraryBuilder` class. I'm not too familiar with phpspec though.

So the biggest improvement here is the addition of the `LibraryBuilder` class which has some methods that can be used to manipulate the search results. all API fetches will go through the `fetch()` method and put the results in a `data` property. The api fetch doesn't happen until the trigger methods like `all()`. This will allow a lot of the methods that manipulate the searches like `chunk()`, `limit()` etc to be stacked before the final api is fetched.

an example would be:

```php
$books->volumes->search('book title')->limit(20)->chunk(5)->all();
$books->volumes->find('GJS6HHS87S');
```

even the volume methods are wrappers around the LibraryBuilder class
`

File: src/LibraryBuilder.php [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+     * @return array
+     * @see chunk()
+     */
+    public function all()
+    {
+        if ($this->chunk) {
+            return $this->data = array_chunk($this->fetch()->items, $this->chunk);
+        }
+        
+        return $this->data = $this->fetch();
```
Comment (josephmancuso) [inline src/LibraryBuilder.php]: this needs to return `$this->fetch()->items` since it will return an array of the `Volume` class

File: spec/VolumesSpec.php [hunk 6]
```diff
@@ -36,34 +37,34 @@ function it_should_provide_search()
...
-        $res->shouldHaveType(\Generator::class);
+        $res->shouldHaveType(LibraryBuilder::class);
     }
 
     function it_should_provide_search_based_lookup()
     {
         $this->init('volumes.json', 'volume.json');
         $res = $this->firstOrNull('isbn:0253324009');
-        $res->shouldHaveType(Volume::class);
+        $res->shouldHaveType(\stdclass::class);
```
Comment (danmichaelo) [inline spec/VolumesSpec.php:47]: Hm, why this change?

File: src/GoogleBooks.php [hunk 7]
```diff
@@ -19,7 +19,7 @@ class GoogleBooks
     /**
      * @var Client
      */
-    protected $http;
+    public $http;
```
Comment (danmichaelo) [inline src/GoogleBooks.php]: Or we could leave it protected and allow it to be injected in the constructor. Something like

```
public function __construct($options = [], Client $http = null)
    {
        $this->http = $http ?: new Client([
            'base_uri' => $this->baseUri,
            'handler' => isset($options['handler']) ? $options['handler'] : null,
        ]);
```

Reminds me that I would like to add http://httplug.io/, but let's keep that out for now :)

File: src/GoogleBooks.php [hunk 8]
```diff
@@ -70,7 +70,7 @@ public function __construct($options = [])
         $this->maxResults = isset($options['maxResults']) ? $options['maxResults'] : null;
     }
 
-    protected function raw($endpoint, $params = [], $method='GET')
+    public function raw($endpoint, $params = [], $method='GET')
```
Comment (danmichaelo) [inline src/GoogleBooks.php:73]: Was this just to make phpspec happy? If so, is there another way?

File: src/GoogleBooks.php [hunk 9]
```diff
@@ -119,24 +118,34 @@ public function getItem($path)
...
+    }
 
+    public function listItems($endpoint, $params = [])
+    {
         $i = 0;
+        $return = [];
         while (true) {
             $n = $i % $this->batchSize;
+            $params['maxResults'] = $this->batchSize;
+
```
Comment (danmichaelo) [inline src/GoogleBooks.php]: Perhaps we should add a `.editorconfig` file? I prefer having "trim_trailing_whitespace: true"

File: src/LibraryBuilder.php [hunk 10]
```diff
@@ -0,0 +1,143 @@
+<?php 
+
+namespace Scriptotek\GoogleBooks;
+
+use GuzzleHttp\Client;
+use Scriptotek\GoogleBooks\GoogleBooks;
+
+class LibraryBuilder
+{
+    private $data;
```
Comment (danmichaelo) [inline src/LibraryBuilder.php]: I usually default to `protected` for reasons like [these](https://stackoverflow.com/a/38876771/489916). I'm not religious about it though, but perhaps we should agree on a standard within the library? Unless you think there's specific reasons for going private here.

File: src/LibraryBuilder.php [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    private function fetch($id = false)
+    {
+        if ($this->limit <= 40) {
+            if ($id !== false) {
+                return $this->client->getItem("volumes/$id");
+            }
+            return $this->client->raw('volumes', $this->getParams());
+        }
+        
+        return $this->fetchAll();
```
Comment (danmichaelo) [inline src/LibraryBuilder.php:65]: Hm, I specify `->limit(50)`, I wouldn't expect to get *all* :)

File: src/Volumes.php [hunk 12]
```diff
@@ -2,10 +2,12 @@
 
 namespace Scriptotek\GoogleBooks;
 
+use Scriptotek\GoogleBooks\LibraryBuilder;
 
 class Volumes
 {
     protected $client;
+    protected $chunk = false;
```
Comment (danmichaelo) [inline src/Volumes.php]: Not in use?

File: src/Volumes.php [hunk 13]
```diff
@@ -14,30 +16,40 @@ public function __construct(GoogleBooks $client)
...
+     * @return \Generator|Volume || array
      */
     public function search($query)
     {
+        return (new LibraryBuilder($this->client))->search($query);
+    }
+
+    /**
+     * @param $query
+     * @return \Generator|Volume
```
Comment (danmichaelo) [inline src/Volumes.php]: No longer Generator (btw. phpdoc doesn't have good support for documenting generators as you see)

**Review by danmichaelo — CHANGES_REQUESTED** [review]
Overall I love the approach!

File: spec/VolumesSpec.php (hunk 6 above)
Comment (josephmancuso) [reply spec/VolumesSpec.php:47]: so a the `Volume` class uses `LibraryBuilder` methods which only care about the data that is fetched. I can make it return a volume class but then it might look more like:

```php
$books->volumes->search('hey')->all()->data;
```

the all() just cares about the data property which returns an standard object

File: src/GoogleBooks.php (hunk 7 above)
Comment (josephmancuso) [reply src/GoogleBooks.php]: hmm . I don't remember why I even made it public. It might have been a testing thing. I put it back to private and the methods and tests work. Do you know why I might have made it public? lol

File: src/GoogleBooks.php (hunk 8 above)
Comment (josephmancuso) [reply src/GoogleBooks.php:73]: no, this is so I could keep the API calls inside the `GoogleBooks` class and access them through the `fetch()` method inside the `LibraryBuilder` class.

`LibraryBuilder` class takes a `GoogleBooks` dependency (client) and then can access anything it needs from the `LibraryBuilder`.

This way we keep communication with GoogleBooks inside the GoogleBooks class and the data manipulation methods inside the `LibraryBuilder` class

File: src/GoogleBooks.php (hunk 9 above)
Comment (josephmancuso) [reply src/GoogleBooks.php]: haven't used a `.editorconfig` file. can you elaborate more?

File: src/LibraryBuilder.php (hunk 11 above)
Comment (josephmancuso) [reply src/LibraryBuilder.php:65]: true. The limits above 40 actually don't work as intended here. Since the API has a 40 result limit, the trick here is to keep calling the api and moving the startindex. This works but it doesn't stop at 50 results, it keeps going and gets all the results. I'll prob open an issue and we can get to this some other way.

File: src/LibraryBuilder.php (hunk 10 above)
Comment (josephmancuso) [reply src/LibraryBuilder.php]: ok we can go with protected unless needed to be private

File: src/Volumes.php (hunk 13 above)
Comment (josephmancuso) [reply src/Volumes.php]: ill change this

**Review by josephmancuso — COMMENTED** [review]
I'll make some changes. I've commented on a few things explaining my reasoning

File: spec/VolumesSpec.php (hunk 6 above)
Comment (danmichaelo) [reply spec/VolumesSpec.php:47]: Yeah, but it makes more sense that the`Volumes` class returns `Volume`, doesn't it? What about injecting the desired class into LibraryBuilder, smtn like `new LibraryBuilder($this->client, Volume::class))`?

File: src/GoogleBooks.php (hunk 8 above)
Comment (danmichaelo) [reply src/GoogleBooks.php:73]: Ah, yeah, didn't see that. Makes sense.

File: src/LibraryBuilder.php (hunk 11 above)
Comment (danmichaelo) [reply src/LibraryBuilder.php:65]: Yeah. Perhaps just throw an exception if limit is set to more than 40 for now, so we don't forget?

File: src/GoogleBooks.php (hunk 9 above)
Comment (danmichaelo) [reply src/GoogleBooks.php]: See http://editorconfig.org/ . Quite useful to keep things consistent when people use different editors (and platforms). I suggest something like this:

```
root = true

[*]
end_of_line = lf
insert_final_newline = true

[*.{php}]
charset = utf-8
indent_style = space
indent_size = 4
```

File: spec/VolumesSpec.php (hunk 6 above)
Comment (josephmancuso) [reply spec/VolumesSpec.php:47]: out of curiosity, what would be the benefit of returning a `Volume` class? are you trying to gain access to additional methods after the api call is fetched?

Because if i make an `all()` method then i'm going to be excepting data i can throw in a foreach

Comment (danmichaelo) [reply spec/VolumesSpec.php:47]: Yeah, convenience methods such as `$volume->getCover()`. We could still make the class traversable so you can use foreach. See #5 (based on master, but shouldn't have many conflicts with your branch)

**Comment by danmichaelo** [comment]
Suggestion: Let's process #5 first. If you're happy with that approach, we can rebase this branch on top of that.
